onnxruntime/python/onnxruntime_pybind_state.cc

```diff
@@ -188,6 +195,70 @@ namespace py = pybind11;
 using namespace onnxruntime;
 using namespace onnxruntime::logging;
 
+// Custom op section starts
+static Env& platform_env = Env::Default();
+
+CustomOpLibrary::CustomOpLibrary(const char* library_path, OrtSessionOptions& ort_so) {
+  {
+    OrtPybindThrowIfError(platform_env.LoadDynamicLibrary(library_path, &library_handle_));
+
+    if (!library_handle_)
+      throw std::runtime_error("RegisterCustomOpsLibrary: Failed to load library");
+
+    OrtStatus*(ORT_API_CALL * RegisterCustomOps)(OrtSessionOptions * options, const OrtApiBase* api);
+
+    OrtPybindThrowIfError(platform_env.GetSymbolFromLibrary(library_handle_, "RegisterCustomOps", (void**)&RegisterCustomOps));
+
+    if (!RegisterCustomOps)
+      throw std::runtime_error("RegisterCustomOpsLibrary: Entry point RegisterCustomOps not found in library");
+
+    auto* status = RegisterCustomOps(&ort_so, OrtGetApiBase());
+
+    if (status) {
+      // A non-nullptr indicates status indicates some error
+
+      // TODO: How to handle unload failure ?
+      // Currently we ignore the returned status assuming it is successful
+      platform_env.UnloadDynamicLibrary(library_handle_);
+
+      // Construct error message string
+      std::string error_string = status->msg;
+
+      // Free status
+      ::free(status);
+
+      // Throw
+      throw std::runtime_error(error_string);
+    }
+
+    // No status to free if it is a nullptr
+  }
+}
+
+// Unload the library when the destructor is triggered
+CustomOpLibrary::~CustomOpLibrary() {
+  UnloadLibrary();
+}
+
+// Logic to unload the library
+void CustomOpLibrary::UnloadLibrary() {
+  auto status = platform_env.UnloadDynamicLibrary(library_handle_);
+
+  if (!status.IsOK()) {
+    throw std::runtime_error("Unable to unload a custom op shared library");
+  }
+}
+
+void CustomOpLibraries::AddLibrary(std::unique_ptr<CustomOpLibrary> custom_op_library) {
+  // Guard against multiple session options being created in multiple threads referencing
+  // custom op libraries of their own - basically guard against unsynchronized updates to the
+  // container holding the libraries
+  std::lock_guard<std::mutex> lock(mutex_);
```

Comment:
If the CustomOpLibraries obj is contained inside PySessionOptions, you won't need this mutex.

Reply:
Mutex goes away in the latest design